### DISCUSSION TITLE
SvelteKit based project unable to specify includeFiles to include in the serverless runtime

### DISCUSSION THREAD

**Discussion by louis-byers** [vercel/community#2737] [Help]
### Problem Description

The main project I am working on has an issue with a dependency of a dependency. To illustrate the core issue I created a much simpler project that just does the minimum to trigger the issue.

It can be found at: https://github.com/louis-byers/vercel-node-modules-file-test

The core issue is that the dependency is a node module that wants to load up some jsonld files it has. But they are not in the serverless runtime. `Unhandled Promise Rejection 	{"errorType":"Runtime.UnhandledPromiseRejection","errorMessage":"Error: ENOENT: no such file or directory, open '/var/task/node_modules/security-context/contexts/security-v1.jsonld'",....`

Reading up I was kind of surprised the static analysis didn't find the file to include (src here: https://github.com/w3c-ccg/security-vocab/blob/main/js/index.js)

So after a bit of research at https://github.com/vercel/next.js/discussions/14807 and https://github.com/orgs/vercel/discussions/1180 and https://vercel.com/docs/concepts/functions/serverless-functions/runtimes#including-additional-files and https://vercel.com/docs/concepts/projects/project-configuration#value-definition

I followed instructions to try and force the include. (that `vercel.json` file is checked into the repo above) I tried a few glob patterns but got the following errors:

```
[11:08:14.946] Running build in Washington, D.C., USA (East) – iad1
[11:08:15.007] Cloning github.com/louis-byers/vercel-node-modules-file-test (Branch: main, Commit: 4f72287)
[11:08:15.230] Cloning completed: 222.411ms
[11:08:16.288] Restored build cache
[11:08:16.315] Running "vercel build"
[11:08:16.799] Vercel CLI 29.3.6
[11:08:16.928] Error: The pattern "api/**/*" defined in `functions` doesn't match any Serverless Functions inside the `api` directory.
[11:08:16.929] Learn More: https://vercel.link/unmatched-function-pattern
[11:08:16.962] unused_function: The pattern "api/**/*" defined in `functions` doesn't match any Serverless Functions inside the `api` directory.
```

and

```
[11:09:53.573] Running build in Washington, D.C., USA (East) – iad1
[11:09:53.640] Cloning github.com/louis-byers/vercel-node-modules-file-test (Branch: main, Commit: 271401a)
[11:09:53.895] Cloning completed: 254.24ms
[11:09:54.959] Restored build cache
[11:09:54.988] Running "vercel build"
[11:09:55.500] Vercel CLI 29.3.6
[11:09:55.648] Error: The pattern "**/*" defined in `functions` doesn't match any Serverless Functions inside the `api` directory.
[11:09:55.648] Learn More: https://vercel.link/unmatched-function-pattern
[11:09:55.760] unused_function: The pattern "**/*" defined in `functions` doesn't match any Serverless Functions inside the `api` directory.
```

As it is not a direct dependency of my real project, it is more difficult to do this: https://vercel.com/guides/why-does-my-serverless-function-work-locally-but-not-when-deployed because it would take some coordination with the library authors.

I would love any help anyone can offer or other ideas to try to get this working in the deployed environment.

Thanks so much!!

### Example

https://github.com/louis-byers/vercel-node-modules-file-test

### Steps to Reproduce

1. Create a sveltekit project
2. On a server rendered page e.g. `+page.server.js` include the node module `security-context` and do anything with it. I just printed out the contexts.
3. Look at vercel deployment
4. :boom:

**Comment by louis-byers** [comment]
someone with a similar problem: https://stackoverflow.com/questions/74769949/vercel-cannot-find-static-files-from-node-modules-in-nextjs

**Comment by louis-byers** [comment]
Full stack trace of error in case that helps some:

```
Unhandled Promise Rejection 	{"errorType":"Runtime.UnhandledPromiseRejection","errorMessage":"Error: ENOENT: no such file or directory, open '/var/task/node_modules/security-context/contexts/security-v1.jsonld'","reason":{"errorType":"Error","errorMessage":"ENOENT: no such file or directory, open '/var/task/node_modules/security-context/contexts/security-v1.jsonld'","code":"ENOENT","errno":-2,"syscall":"open","path":"/var/task/node_modules/security-context/contexts/security-v1.jsonld","stack":["Error: ENOENT: no such file or directory, open '/var/task/node_modules/security-context/contexts/security-v1.jsonld'","    at Object.openSync (node:fs:601:3)","    at Object.readFileSync (node:fs:469:35)","    at _read (/var/task/node_modules/security-context/js/index.js:12:8)","    at Object.<anonymous> (/var/task/node_modules/security-context/js/index.js:19:3)","    at Module._compile (node:internal/modules/cjs/loader:1254:14)","    at Module._extensions..js (node:internal/modules/cjs/loader:1308:10)","    at Module.load (node:internal/modules/cjs/loader:1117:32)","    at Module._load (node:internal/modules/cjs/loader:958:12)","    at ModuleWrap.<anonymous> (node:internal/modules/esm/translators:169:29)","    at ModuleJob.run (node:internal/modules/esm/module_job:194:25)"]},"promise":{},"stack":["Runtime.UnhandledPromiseRejection: Error: ENOENT: no such file or directory, open '/var/task/node_modules/security-context/contexts/security-v1.jsonld'","    at process.<anonymous> (file:///var/runtime/index.mjs:1189:17)","    at process.emit (node:events:525:35)","    at emit (node:internal/process/promises:149:20)","    at processPromiseRejections (node:internal/process/promises:283:27)","    at process.processTicksAndRejections (node:internal/process/task_queues:96:32)"]}
Unknown application error occurred
Runtime.Unknown
```

**Comment by louis-byers** [comment]
We managed to find another version of our dependency that no longer relies on the security-context library that was problematic for the Vercel deployment environment.

**Comment by fabiancook** [comment]
I was running into this issue too

![image](https://github.com/vercel/community/assets/4337080/e8bf765e-7904-4015-b4ab-6d3b840c258b)

It appears that vercel requires strong named references from one file to another for the file. I'm not sure how they do this, but it caused some issues around using esm where the files were being served from an importmap. 

I figured this out by listing, within the deployment, the contents of the folder that was missing files. 

![image](https://github.com/vercel/community/assets/4337080/a2226e41-7244-4621-8241-93ff516e7d5e)

The files were definitely included in the npm module, but apart from the json files, the js files were not included. 

![image](https://github.com/vercel/community/assets/4337080/ef354394-3984-42a3-92af-f960099f81e6)
![image](https://github.com/vercel/community/assets/4337080/f9f0a2af-aa16-4d15-b233-0a166b0816c6)



I added some more code to my post build script that generates the importmap, that creates another file with all the references. 

https://github.com/socialbaking/karma/blob/78066dd07ce22ffcbdb3a52f62335278ea1fadf1/scripts/importmap.js#L92-L101 

This made sure I knew there were references. 

Including this one change, had the files pop up in the deployment. 

Heres the file of all the references to the missing files: 

![image](https://github.com/vercel/community/assets/4337080/32fc21d2-a0df-459c-8751-257d25019347)

I then included a reference to this file in my main handler function. 

Then, after deployment, the files were now included. 

![image](https://github.com/vercel/community/assets/4337080/24cefa2f-5dd2-4509-8f3e-465fbc15c57e)

And the file became available :)

![image](https://github.com/vercel/community/assets/4337080/e366e249-8cb9-4299-838e-0f7914bff86f)


Before this commit it was broken, after it worked, so confirms it was a referencing issue:

https://github.com/socialbaking/karma/commit/b9add6be34ed70e49410f607c2581daff249106a

Vercel should ideally provide a way to define these kinds of paths.

**Reply by fabiancook** [reply]
I had the same thing happen now when I changed some file references to my public files. 

Making another change to include these references, made files that were returning a 404, be included in the deployment, and be resolved https://github.com/socialbaking/karma/issues/116 

Import map references:

```typescript
export async function importReferences() { 
    try {
    await import("./imports/react/jsx-runtime.js");
await import("./imports/react.js");
await import("./imports/react-dom.js");
await import("./imports/react-dom/client.js");
await import("./imports/fuzzy-search.js");
await import("./imports/is.js");
await import("./imports/@simplewebauthn/browser.js");
await import("./imports/@authsignal/browser.js");
    } catch {
    
    }
    }
```

Other public static file references:

```typescript
export async function staticReferences() { 
    try {
    await import("../public/example-1.svg");
await import("../public/example-2.svg");
await import("../public/example-3.svg");
await import("../public/named-watermark.svg");
await import("../public/watermark.png");
await import("../public/watermark.svg");
await import("./karma/react/client/pages/index.js");
await import("./karma/react/client/is.js");
await import("./karma/react/client/utils.js");
    } catch {
    
    }
    }
```

Originally I had included the references as direct exports, but I found that if you use references inside of `await import` this also counts, which lets me list a bunch of references without ever including them for use, they are only there for vercel to see I'm going to utilise them. 

![image](https://github.com/vercel/community/assets/4337080/b0b965b4-06c3-4f69-bdbd-99bb4f4a912c)

![image](https://github.com/vercel/community/assets/4337080/69506b67-f14a-41a7-abba-0edc303100de)

The difference between the above two requests is including the references...
